### PR TITLE
Fixing a small, but significant, bug with MOON.

### PR DESCRIPTION
**Pull request by emersodb**
# PR Type
Fix

# Short Description

Fixing a small bug in the MOON implementation in line with Sana's fix. Adding some tests to ensure the fix is properly functioning.

# Tests Added

Some tests to make sure the loss calculations etc. function as expected given the change.

### CODE REVIEW

File: fl4health/clients/moon_client.py [hunk 1]
```diff
@@ -89,14 +91,17 @@ def get_contrastive_loss(
 
         return self.ce_criterion(logits, labels)
 
-    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+    def get_parameters(self, config: Config) -> NDArrays:
         assert isinstance(self.model, MoonModel)
-        # Save the parameters of the old local model
+        # Save the parameters of the old LOCAL model
```
Comment (lotif) [inline fl4health/clients/moon_client.py:96]: Why is the saving being done on get now instead of set?

**Review by lotif — COMMENTED** [review]
Just one question

File: fl4health/clients/moon_client.py (hunk 1 above)
Comment (emersodb) [reply fl4health/clients/moon_client.py:96]: That's a good question and the crux of the issue that Sana identified. `set_parameters` is called to start both the `fit` and `eval` stages. The eval stage happens after aggregation. So in the simplified case, when `old_models_list` is supposed to have size 1, the `old_local_model` that MOON needs to compute features is replaced with the global model produced by aggregation. `get_parameters` is only called at the end of each `fit`, which means that old_models_list will always have the most recent model produced by **local** training, which is what we want.

Comment (lotif) [reply fl4health/clients/moon_client.py:96]: Just for the record, I think this would be better if moved to its own function, but as per offline conversation, this will be done in a follow up ticket (https://app.clickup.com/t/86879vn75)

**Review by lotif — APPROVED** [review]

File: tests/clients/test_moon_client.py [hunk 2]
```diff
@@ -20,12 +20,64 @@ def test_setting_parameters(get_client: MoonClient) -> None:  # noqa
...
 
+    # Global model parameters should match those of the new_params
+    global_model_params = [val.cpu().numpy() for _, val in moon_client.global_model.state_dict().items()]
+    # Make sure the MOON model parameters are equal to the global model parameters
+    new_moon_model_params = [val.cpu().numpy() for _, val in moon_client.model.state_dict().items()]
+
+    for i in range(len(params)):
+        assert (new_params[i] == global_model_params[i]).all()
+        assert (new_params[i] == new_moon_model_params[i]).all()
+
```
Comment (sanaAyrml) [inline tests/clients/test_moon_client.py:31]: Maybe we can add an assertion here to check whether global_model and old_models_list[-1] is different. (That's what I figured out the bug from)

**Comment by sanaAyrml** [comment]
Changes seems good to me, Thanks for applying changes.

**Review by sanaAyrml — APPROVED** [review]
Changes seems good to me, Thanks for applying changes.

File: tests/clients/test_moon_client.py (hunk 2 above)
Comment (emersodb) [reply tests/clients/test_moon_client.py:31]: So I can't do it in this part, because I assume that `len(moon_client.old_models_list) == 0` just below. However, below, on both lines 59 and 78, there is an assertion that they are different. I'll add in setting the model twice in the last check, just to be certain.

**Review by fatemetkl — APPROVED** [review]
